showkase-component-impressions/src/main/java/com/vinaygaba/showkase_component_impressions/ComponentImpressions.kt

```diff
@@ -0,0 +1,145 @@
+ * visibility of a given composable function that uses the visibilityEvents Modifier.
+ */
+data class ShowkaseVisibilityEvent<T>(
+    val key: T,
+    val visibilityPercentage: Float,
+    val boundsInWindow: Rect
+)
+
+/**
+ * Use this modifier to get visibility events for a given Composable. It emits visibility events
+ * when the composable is added to the composition (visible), when its removed from the
+ * composition(invisible), when the activity is backgrounded(visible) and when the activity is
+ * foregrounded(visible). In addition,
+ *
+ * @param key Unique identifier for a given composable function that you use this modifier on
+ * @param onVisibilityChanged Callback that's called when the visibility of a composable function
+ * changes. This event has information about the key of the composable, the percentage of composable
+ * that's visible on the screen and the bounds of the composable.
+ */
+@OptIn(FlowPreview::class)
+fun <T> Modifier.visibilityEvents(
+    key: T,
+    onVisibilityChanged: (ShowkaseVisibilityEvent<T>) -> Unit,
+) = composed {
+    val view = LocalView.current
+    val lifecycle = LocalLifecycleOwner.current.lifecycle
+    val scope = rememberCoroutineScope()
+    val visibilityEvent by rememberUpdatedState(newValue = onVisibilityChanged)
+    val impressionCollector = remember(key) { ImpressionCollector<T>(scope, visibilityEvent) }
+    var visibilityMetadata: VisibilityMetadata? by remember { mutableStateOf(null) }
+
+    registerLifecycleImpressionEvents(key, lifecycle, impressionCollector, visibilityMetadata)
+    registerDisposeImpressionEvents(key, impressionCollector, lifecycle)
+    collectImpressionEvents(key, impressionCollector, visibilityEvent)
+
+    onGloballyPositioned { layoutCoordinates ->
+        visibilityMetadata = layoutCoordinates.visibilityPercentage(view = view)
+        impressionCollector.onLayoutCoordinatesChanged(key, visibilityMetadata!!)
+    }
+}
+
+/**
+ * Used for handling the use case where a composable function is not in composition anymore i.e is
+ * invisible.
+ */
+@SuppressLint("ComposableNaming")
+@Composable
+private fun <T> registerDisposeImpressionEvents(
+    key: T,
+    impressionCollector: ImpressionCollector<T>,
+    lifecycle: Lifecycle
+) {
+    DisposableEffect(key, lifecycle) {
+        onDispose {
+            impressionCollector.publishDisposeEvent(
+                key,
+                defaultVisibilityMetadata,
+            )
+        }
+    }
+}
+
+/**
+ * Collects the visibility change events from the internal coroutine flow and invokes the
+ * onVisibilityEvent that the user passed.
+ */
+@SuppressLint("ComposableNaming")
+@FlowPreview
+@Composable
+private fun <T> collectImpressionEvents(
+    key: T,
+    impressionCollector: ImpressionCollector<T>,
+    onVisibilityEvent: (ShowkaseVisibilityEvent<T>) -> Unit
+) {
+    LaunchedEffect(key) {
+        impressionCollector.impressionEvents
+            .collect { impressionData ->
```

Comment:
it feels a bit weird to me to have the impression collector collect the events, but then expose them as a flow for this code to recollect to do the actual callback. Could you instantiate the impression collector with the coroutine scope and callback lambda and have that class manage all of this instead? that might make recomposition more efficient too

Reply:
I think this pattern is followed in multiple places in the Compose codebase. This could've been a compose state object too but keeping it as a flow object gives me access to the operators that I was hoping to use.

Reply:
Wondering what you had in mind though. Could you elaborate a bit more @elihart